File: WindowsForm_sample/CustomerServiceTest.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Xunit;
using Testcontainers.PostgreSql;

namespace WindowsForm_sample
{
    public class CustomerServiceTest : IAsyncLifetime
    {
        public readonly PostgreSqlContainer _postgres = new PostgreSqlBuilder()
            .WithImage("postgres:15-alpine").Build();


        public Task InitializeAsync()
        {
            return _postgres.StartAsync();
        }

        public Task DisposeAsync()
        {
            return _postgres.DisposeAsync().AsTask();
        }

        [Fact]
        public void ShouldReturnTwoCustomers()
        {
            // Given  

            var customerService = new CustomerService(new DbConnectionProvider(_postgres.GetConnectionString()));

            // When  
            customerService.Create(new Customer(1, "George"));
            customerService.Create(new Customer(2, "John"));
            var customers = customerService.GetCustomers();

            // Then  
            Assert.Equal(2, customers.Count());
        }
    }
}

File: xunitsample/test/CustomerServiceTest.vb
Imports System.Threading.Tasks
Imports Xunit
Imports Testcontainers.PostgreSql
Imports Testcontainers.MsSql


Public Class CustomerServiceTest
    Implements IAsyncLifetime

    Private ReadOnly _sqlServer As MsSqlContainer = New MsSqlBuilder().Build()

    Public Function InitializeAsync() As Task Implements IAsyncLifetime.InitializeAsync
        Return _sqlServer.StartAsync()
    End Function

    Public Function DisposeAsync() As Task Implements IAsyncLifetime.DisposeAsync
        Return _sqlServer.DisposeAsync().AsTask()
    End Function

    <Fact>
    Public Sub ShouldReturnTwoCustomers()

        ' Given
        Dim provider As New DbConnectionProvider(_sqlServer.GetConnectionString())
        Dim customerService As New CustomerService(provider)

        ' When
        customerService.Create(New Customer(1, "George"))
        customerService.Create(New Customer(2, "John"))
        Dim customers = customerService.GetCustomers()

        ' Then
        Assert.Equal(2, customers.Count())
    End Sub

End Class

